### PR TITLE
Don't require style sharing cache invalidation for :dir

### PR DESCRIPTION
**Pull request by bzbarsky**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because we have no way to test performance optimizations, sadly.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16875)

### CODE REVIEW

File: components/style/gecko/selector_parser.rs [hunk 1]
```diff
@@ -346,7 +346,12 @@ impl NonTSPseudoClass {
 
     /// Returns true if the given pseudoclass should trigger style sharing cache revalidation.
     pub fn needs_cache_revalidation(&self) -> bool {
-        self.state_flag().is_empty() &&
+        // :dir() depends on state only, but doesn't use state_flag because its
+        // semantics don't quite match.  Nevertheless, it doesn't need cache
+        // revalidation, because we already compare states for elements and
+        // candidates.
+        (self.state_flag().is_empty() ||
+         matches!(*self, NonTSPseudoClass::Dir(_))) &&
```
Comment (emilio) [inline components/style/gecko/selector_parser.rs]: Hmm... Is this bit right? Shouldn't this be:

```rust
self.state_flag().is_empty() && !matches(*self, NonTSPseudoClass::Dir(_) | NonTSPseudoClass::MozAny(_))
```

Comment (bzbarsky) [inline components/style/gecko/selector_parser.rs]: And this is why we have code review.  You're absolutely right.

**Comment by bzbarsky** [comment]
@emilio Should be better now.  ;)

**Comment by emilio** [comment]
Indeed it is :)

@bors-servo r+

**Comment by bholley** [comment]
Let's double-check that that's real.

@bors-servo try

**Comment by emilio** [comment]
@bors-servo retry

 * Wat

**Comment by nox** [comment]
@bors-servo p=5

Some other patches depend on that.

**Pull request merged**